# AtRTD2-DE-DAS-DTU-pipeline
The R source code of differential expression (DE), differential alternative splicing (DAS) and differential transcript usage (DTU) analysis pipeline for the paper “Rapid and dynamic alternative splicing impacts the Arabidopsis cold response” is in the file "DE and DAS analysis R codes for AtRTD2 RNA-seq.Rmd" 



<h2 id="fig"> </h2>

![](https://github.com/wyguo/AtRTD2-DE-DAS-DTU-pipeline/blob/master/fig/pipeline.png)

# Reference
1. Calixto et al. (submitted) Rapid and dynamic alternative splicing impacts the Arabidopsis cold response.
2. Law,C. et al. (2016) RNA-seq analysis is easy as 1-2-3 with limma, Glimma and edgeR. F1000Research, 5.
3. Lu,X. et al. (2017) RNA-seq Analysis of Cold and Drought Responsive Transcriptomes of Zea mays ssp. mexicana L. Front. Plant Sci., 8, 1–15.
4. Risso,D. et al. (2014) Normalization of RNA-seq data using factor analysis of control genes or samples. Nat. Biotechnol., 32, 896–902.
5. Ritchie,M.E. et al. (2015) limma powers differential expression analyses for RNA-sequencing and microarray studies. Nucleic Acids Res, 43, e47.
6. Robinson,M.D. et al. (2010) edgeR: a Bioconductor package for differential expression analysis of digital gene expression data. Bioinformatics, 26, 139–140.
7. Soneson,C. et al. (2015) Differential analyses for RNA-seq: transcript-level estimates improve gene-level inferences. F1000Res, 4, 1521.

