---
title: "DE DAS and DTU analysis"
date: '`r format(Sys.time(), "%d %B, %Y")`'
header-includes: \usepackage{caption}
output:
  html_document:
    fig_caption: yes
    highlight: textmate
    theme: cerulean
    toc: yes
    toc_depth: 4
    toc_float:
      collapsed: no
      smooth_scroll: yes
  pdf_document:
    fig_caption: yes
    toc: yes
    toc_depth: '4'
  word_document:
    fig_caption: yes
    toc: no
    toc_depth: '4'
params:
  doc.type: html_document
  wkd: NULL

---

```{r setup, include=FALSE,eval=T}
knitr::opts_chunk$set(echo = F,eval = T)
options(stringsAsFactors = F)
################################################
##--load library
require(readr)
require(eulerr)
require(gridExtra)
require(grid)
require(Gmisc)
require(dplyr)

wkd <- params$wkd

# setwd("D:/PhD project/R projects for PhD######/test round 2018/Immediate dataset")
para <- data.frame(read.csv('result/Parameter summary.csv'))
colnames(para) <- c('Step','Description','Parameter')
data.info <- suppressMessages(read_csv('result/data.info.csv'))
DDD.number <- suppressMessages(read_csv('result/DE DAS DTU numbers.csv'))
DEvsDAS <- suppressMessages(read_csv('result/DE vs DAS gene number.csv'))
DEvsDTU <- suppressMessages(read_csv('result/DE vs DTU transcript number.csv'))
load('data/contrast.RData')
load('data/samples.RData')
load('data/samples_new.RData')
load('data/DE_genes.RData')
load('data/DAS_genes.RData')
load('data/DE_trans.RData')
load('data/DTU_trans.RData')
load('data/target_high.RData')

sourceDir <- function(path, trace = TRUE, ...) {
  for (nm in list.files(path, pattern = "[.][RrSsQq]$")) {
    #if(trace) cat(nm,":")
    source(file.path(path, nm), ...)
    #if(trace) cat("/n")
  }
}
sourceDir('D:/PhD project/R projects for PhD######/R function in general')

# sourceDir('E:/DDD/R function in general')

fig.caption <- function(cap){
  if(params$doc.type=='html_document' | params$doc.type=='word_document')
    paste0('Figure: ',cap) else cap
}


```

## Overview
Working directory for the analysis:
```{r}
getwd()
knitr::kable(para,format = "markdown") 
```

Table: Overview of parameters used for the DE, DAS and DTU analysis.

## Experimental design
```{r,results='asis'}
cat('  \nConditions ( n=',length(unique(samples$condition)),'):',unique(samples$condition))
cat('  \nBiological replicate number:',length(unique(samples$brep)))
cat('  \nSequencing replicate number:',length(unique(samples$srep)))
cat('  \nContrast groups for expression comparisons:')
x <- do.call(rbind,strsplit(contrast,'-'))
colnames(x) <- c('Treatment','Control')
knitr::kable(x,format = "markdown") 
```

Table: Contrast groups.

## Expression data generation
```{r,results='asis'}
cat('The',as.character(para$Parameter[para$Description=='tximport method']),'method in [tximport](https://bioconductor.org/packages/release/bioc/html/tximport.html) R package (Soneson, Love, and Robinson 2016) was used to generate gene and transcript expression.')

```

## Data pre-processing
![](https://github.com/wyguo/AtRTD2-DE-DAS-DTU-pipeline/blob/master/fig/data_preprocessing.png?raw=true)


### Step 1: Merge sequencing replicates
```{r,results='asis'}
if(as.character(para$Parameter[para$Description=='Merge sequencing replicates'])=='Yes'){
  cat('The sequencing replciates of the dataset were merged.') } else {
    cat('The dataset has no sequencing replicates.')
  }
```

### Step 2: Filter low expression
The cut-offs to filter the transcripts were determined by the mean-variance trend plots (Law et al. 2014).
```{r,results='asis'}
cat('\n- An expressed transcript must have $\\geq$',
    para$Parameter[para$Description=='Sample number for CPM cut-off'],'out of',length(samples_new$condition),
    'samples with CPM $\\geq$',para$Parameter[para$Description=='Low expression CPM cut-off'])
cat('\n- An expressed gene must have at least one expressed transcript.\n')
```

```{r,echo=FALSE, out.width = '100%',fig.cap=fig.caption('Transcript mean-variance trend')}
if(!file.exists('figure/Transcript mean-variance trend.png')){
  cat('The file "Transcript mean-variance trend.png" is not in the figure folder.')
} else {
knitr::include_graphics("figure/Transcript mean-variance trend.png")
}
```

```{r,echo=FALSE, out.width = '100%',fig.cap=fig.caption('Gene mean-variance trend')}
if(!file.exists('figure/Gene mean-variance trend.png')){
  cat('The file "figure/Gene mean-variance trend.png" is not in the figure folder.')
} else {
knitr::include_graphics("figure/Gene mean-variance trend.png")
}
```
<hr>
```{r,results='asis'}
knitr::kable(data.info,format = "markdown")
```

Table: Data inforamtion after low expression filters.


### Step 3: PCA plots

#### PCA plot of biological replicates
Use the PCA plot to check batch effects between biological replicates

```{r,echo=FALSE, out.width = '100%',fig.cap=fig.caption('Transcript level PCA plot of biological replicates')}
if(!file.exists('figure/Transcript PCA_Bio-reps.png')){
  cat('The file "Transcript PCA_Bio-reps.png" is not in the figure folder.')
} else {
knitr::include_graphics("figure/Transcript PCA_Bio-reps.png")
}
```

```{r,echo=FALSE, out.width = '100%',fig.cap=fig.caption('Gene level PCA plot of biological replicates')}
if(!file.exists('figure/Gene PCA_Bio-reps.png')){
  cat('The file "figure/Gene PCA_Bio-reps.png" is not in the figure folder.')
} else {
knitr::include_graphics("figure/Gene PCA_Bio-reps.png")
}
```

```{r,results='asis'}
if(para$Parameter[para$Description=='Batch effect estimation']=='Yes'){
  cat('#### Estimate batch effects (YES)')
  cat('  \nThe data has batch effects. The',para$Parameter[para$Description=='Batch effect estimation method'],
      'method in [RUVSeq](https://bioconductor.org/packages/release/bioc/html/RUVSeq.html) R package (Risso et al. 2014)
      was used to estimate the transcript and gene level batch effect terms between biological replicates, which were 
      incorporated as an additional factor in the linear regression model for expression change analysis.
      The package also generate read counts where the batch effects were removed from the dataset. To avoid data over-correction, the batch removed
      datasets were only used to make the PCA plots rahter than the downstream expression comparision analysis.')
} else {
  cat('#### Estimate batch effects (NO)')
  cat('  \nThe data has no batch effects')
}
```

```{r,echo=FALSE, out.width = '100%',fig.cap=fig.caption('Transcript level PCA plot of biological replicates, batch effects were removed')}
if(!file.exists('figure/Transcript PCA batch effect removed_Bio-reps.png')){
  cat('The file "figure/Transcript PCA batch effect removed_Bio-reps.png" is not in the figure folder.')
} else {
knitr::include_graphics("figure/Transcript PCA batch effect removed_Bio-reps.png")
}
```

```{r,echo=FALSE, out.width = '100%',fig.cap=fig.caption('Gene level PCA plot of biological replicates, batch effects were removed')}
if(!file.exists('figure/Gene PCA batch effect removed_Bio-reps.png')){
  cat('The file "figure/Gene PCA batch effect removed_Bio-reps.png" is not in the figure folder.')
} else {
knitr::include_graphics("figure/Gene PCA batch effect removed_Bio-reps.png")
}
```

#### PCA plot of average expression
To visualise the data quality and variance between conditions, the average read counts of biological replicates (batch effects were removed if exist) were used to make the PCA plot.

```{r,echo=FALSE, out.width = '100%',fig.cap=fig.caption('Transcript level PCA plot of average expression.')}
if(!file.exists('figure/Transcript PCA_Average expression.png')){
  cat('The file "figure/Transcript PCA_Average expression.png" is not in the figure folder.')
} else {
knitr::include_graphics("figure/Transcript PCA_Average expression.png")
}
```

```{r,echo=FALSE, out.width = '100%',fig.cap=fig.caption('Gene level PCA plot of average expression.')}
if(!file.exists('figure/Gene PCA_Average expression.png')){
  cat('The file "figure/Gene PCA_Average expression.png" is not in the figure folder.')
} else {
knitr::include_graphics("figure/Gene PCA_Average expression.png")
}
```

### Step 4: Data normalisation
```{r,results='asis'}
cat('The',para$Parameter[para$Description=='Normalisation method'], 'method (Robinson, McCarthy, and Smyth 2010)
    and [edgeR](https://bioconductor.org/packages/release/bioc/html/edgeR.html)
    R package (Robinson, McCarthy, and Smyth 2010) were used to normalise the transcript and gene level read counts.')

```

```{r,echo=FALSE, out.width = '100%',fig.cap=fig.caption('Transcript level expression distribution.')}
if(!file.exists('figure/Transcript data distribution.png')){
  cat('The file "figure/Transcript data distribution.png" is not in the figure folder.')
} else {
knitr::include_graphics("figure/Transcript data distribution.png")
}
```

```{r,echo=FALSE, out.width = '100%',fig.cap=fig.caption('Gene level expression distribution.')}
if(!file.exists('figure/Gene data distribution.png')){
  cat('The file "figure/Gene data distribution.png" is not in the figure folder.')
} else {
  knitr::include_graphics("figure/Gene data distribution.png")
}
```

## DE DAS and DTU analysis

### DE DAS and DTU definitions
![](https://github.com/wyguo/AtRTD2-DE-DAS-DTU-pipeline/blob/master/fig/DDD_definition.png?raw=true)

### DE DAS and DTU summary
```{r,results='asis'}
pval.cut <- para$Parameter[para$Description=="Adjusted p-value cut-off"]
lfc.cut <- para$Parameter[para$Description=="Log2 fold change cut-off"]
deltaPS.cut <- para$Parameter[para$Description=="delta PS cut-off"]
pipeline <- para$Parameter[para$Description=="Pipeline"]
cat('#### Methods  \n')
if(pipeline=='limma'){
  cat('\n- Pipeline: ',pipeline,'for DE and',para$Parameter[para$Description=="AS function"],'for DAS and DTU (Smyth et al. 2013)')
} else {
  cat('\n- Pipeline: ',pipeline,'for DE and',para$Parameter[para$Description=="AS function"],'for DAS and DTU (Robinson, McCarthy, and Smyth 2010)')
}

cat('\n- P-value adjusted method: ',para$Parameter[para$Description=="P-value adjust method"])
cat('\n- Adjusted p-value cut-off: ',pval.cut)
cat('\n- Log2 fold change: ',lfc.cut)
cat('\n- deltaPS: ',deltaPS.cut)
cat('\n- Contrast groups: ',paste(contrast,collapse = ','),'  \n  \n')

cat('#### Results  \n')
cat('##### Individual contrast group:  \n')

cat('\n- DE DAS and DTU numbers  \n')
knitr::kable(DDD.number,format = "markdown")
cat('<br>\n')

cat('\n- DE genes compared with DAS genes  \n')
knitr::kable(DEvsDAS,format = "markdown")
cat('<br>\n')

cat('\n- DE transcripts compared with DTU transcripts  \n')
knitr::kable(DEvsDTU,format = "markdown")
cat('<br>\n')

cat('\n- DE transcripts compared with DTU transcripts  \n')
knitr::kable(DEvsDTU,format = "markdown")

```

<br>
<br>

##### Union set of all contrast groups:
```{r,echo=FALSE, out.width = '100%',fig.cap=fig.caption('Union set of DE and DAS genes across all contrast groups.')}
if(!file.exists('figure/Union set DE genes vs DAS genes.png')){
  cat('The file "figure/Union set DE genes vs DAS genes.png" is not in the figure folder.')
} else {
  knitr::include_graphics("figure/Union set DE genes vs DAS genes.png")
}
```

```{r,echo=FALSE, out.width = '100%',fig.cap=fig.caption('Union set of DE and DTU transcripts across all contrast groups.')}
if(!file.exists('figure/Union set DE transcripts vs DTU transcripts.png')){
  cat('The file "figure/Union set DE transcripts vs DTU transcripts.png" is not in the figure folder.')
} else {
  knitr::include_graphics("figure/Union set DE transcripts vs DTU transcripts.png")
}
```

## Heatmap
### DE genes
```{r,echo=FALSE, out.width = '80%',fig.cap=fig.caption('Heatmap of DE genes'),fig.align='center'}
if(!file.exists('figure/Heatmap DE genes.png')){
  cat('The file "Heatmap DE genes.png" is not in the figure folder.')
} else {
knitr::include_graphics('figure/Heatmap DE genes.png')
}
```

### DAS genes
```{r,echo=FALSE, out.width = '80%',fig.cap=fig.caption('Heatmap of DAS genes. The DAS only genes were removed in the plot since they have no significant abundance changes in the contrast groups.'),fig.align='center'}
if(!file.exists('figure/Heatmap DAS genes.png')){
  cat('The file "Heatmap DAS genes.png" is not in the figure folder.')
} else {
knitr::include_graphics('figure/Heatmap DAS genes.png')
}
```

### DE transcript
```{r,echo=FALSE, out.width = '80%',fig.cap=fig.caption('Heatmap of DE transcripts.'),fig.align='center'}
if(!file.exists('figure/Heatmap DE transcripts.png')){
  cat('The file "Heatmap DE transcripts.png" is not in the figure folder.')
} else {
knitr::include_graphics('figure/Heatmap DE transcripts.png')
}
```

### DTU transcripts
```{r,echo=FALSE, out.width = '80%',fig.cap=fig.caption('Heatmap of DTU transcripts. The DTU only transcripts were removed in the plot since they have no significant abundance changes in the contrast groups.'),fig.align='center'}
if(!file.exists('figure/Heatmap DTU transcripts.png')){
  cat('The file "Heatmap DTU transcripts.png" is not in the figure folder.')
} else {
  a='figure/Heatmap DTU transcripts.png'
knitr::include_graphics('figure/Heatmap DTU transcripts.png')
}
```

## GO annotation

```{r,results='asis'}
file.idx <- paste0('figure/',list.files(path = paste0('figure'),pattern = glob2rx('*GO annotation*.png')))
if(length(file.idx)==0){
  cat('GO annotation plot of DE genes is not in the figure folder.')
} else {
  do.call(knitr::include_graphics,list(file.idx))
}
cat('  \nFigure: ',paste(gsub('figure/|.png','',file.idx),collapse = '; '),'  \n<hr>')
```

## References
Law, C W, Y Chen, W Shi, and G K Smyth. 2014. "voom: Precision weights unlock linear model analysis tools for RNA-seq read counts." Genome Biol 15 (2): R29. doi:10.1186/gb-2014-15-2-r29.

Risso, Davide, John Ngai, Terence P. Speed, and Sandrine Dudoit. 2014. "Normalization of RNA-seq data using factor analysis of control genes or samples." Nature Biotechnology 32 (9): 896–902. doi:10.1038/nbt.2931.

Robinson, Mark D, Davis J McCarthy, and Gordon K Smyth. 2010. "edgeR: a Bioconductor package for differential expression analysis of digital gene expression data." Journal Article. Bioinformatics (Oxford, England) 26 (1): 139–40. doi:10.1093/bioinformatics/btp616.

Robinson, Mark D., and Alicia Oshlack. 2010. "A scaling normalization method for differential expression analysis of RNA-seq data." Genome Biology 11 (3): R25. doi:10.1186/gb-2010-11-3-r25.

Smyth, Gordon K, Matthew Ritchie, Natalie Thorne, James Wettenhall, and Wei Shi. 2013. "limma:Linear Models for Microarray Data User’s Guide(Now Including RNA-Seq Data Analysis)." R Manual, 1–123.

Soneson, Charlotte, Michael I Love, and Mark D Robinson. 2016. "Differential analyses for RNA-seq: transcript-level estimates improve gene-level inferences." Journal Article. F1000Research 4: 1521. doi:10.12688/f1000research.7563.2.



